pkg/function/client/client.go

```diff
@@ -79,20 +81,51 @@ func (c *client) MapTFn(ctx context.Context, datum *functionpb.Datum) ([]*functi
 }
 
 // ReduceFn applies a reduce function to a datum stream.
-func (c *client) ReduceFn(ctx context.Context, datumStreamCh <-chan *functionpb.Datum) ([]*functionpb.Datum, error) {
+func (c *client) ReduceFn(ctx context.Context, datumStreamCh <-chan *functionpb.Datum) (*functionpb.DatumList, error) {
+	var g errgroup.Group
+	datumList := &functionpb.DatumList{}
+
 	stream, err := c.grpcClt.ReduceFn(ctx)
 	if err != nil {
 		return nil, fmt.Errorf("failed to execute c.grpcClt.ReduceFn(): %w", err)
 	}
-	for datum := range datumStreamCh {
-		if err := stream.Send(datum); err != nil {
-			return nil, fmt.Errorf("failed to execute stream.Send(%v): %w", datum, err)
+	g.Go(func() error {
+		var sendErr error
+		for datum := range datumStreamCh {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			default:
+				if sendErr = stream.Send(datum); sendErr != nil {
+					return sendErr
+				}
+			}
+		}
+		return stream.CloseSend()
+	})
+
+outputLoop:
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+			var resp *functionpb.DatumList
+			resp, err = stream.Recv()
```

Comment:
Just a note here: I think this stream is going to be busy once the window closes. so that means the http/2 stream will pack enough datum objects to keep the overhead to a minimum. but if this assumption is not going to hold, then this becomes one of the perf bottleneck.